mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java

```diff
@@ -0,0 +1,413 @@
+        message3 = createMessage(benwaInboxMailbox, "Subject: Test3 \n\nBody3\n.\n", BODY_START, new PropertyBuilder());
+        message4 = createMessage(benwaWorkMailbox, "Subject: Test4 \n\nBody4\n.\n", BODY_START, new PropertyBuilder());
+    }
+
+    @After
+    public void tearDown() throws MailboxException {
+        producer.cleanUp();
+    }
+
+    @ContractTest
+    public void findShouldReturnEmptyWhenIdListIsEmpty() throws MailboxException {
+        assertThat(sut.find(ImmutableList.<MessageId> of(), FetchType.Full)).isEmpty();
+    }
+
+    @ContractTest
+    public void findShouldReturnOneMessageWhenIdListContainsOne() throws MailboxException {
+        saveMessages();
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(message1.getMessageId()), FetchType.Full);
+        assertThat(messages).containsOnly(message1);
+    }
+
+    @ContractTest
+    public void findShouldReturnMultipleMessagesWhenIdContainsMultiple() throws MailboxException {
+        saveMessages();
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(message1.getMessageId(), message2.getMessageId(), message3.getMessageId()), FetchType.Full);
+        assertThat(messages).containsOnly(message1, message2, message3);
+    }
+
+    @ContractTest
+    public void findShouldReturnMultipleMessagesWhenIdContainsMultipleInDifferentMailboxes() throws MailboxException {
+        saveMessages();
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(message1.getMessageId(), message4.getMessageId(), message3.getMessageId()), FetchType.Full);
+        assertThat(messages).containsOnly(message1, message4, message3);
+    }
+
+    @ContractTest
+    public void findMailboxesShouldReturnEmptyWhenMessageDoesntExist() throws MailboxException {
+        assertThat(sut.findMailboxes(mapperProvider.generateMessageId())).isEmpty();
+    }
+
+    @ContractTest
+    public void findMailboxesShouldReturnOneMailboxWhenMessageExistsInOneMailbox() throws MailboxException {
+        saveMessages();
+        List<MailboxId> mailboxes = sut.findMailboxes(message1.getMessageId());
+        assertThat(mailboxes).containsOnly(benwaInboxMailbox.getMailboxId());
+    }
+
+    @ContractTest
+    public void findMailboxesShouldReturnTwoMailboxesWhenMessageExistsInTwoMailboxes() throws MailboxException {
+        saveMessages();
+
+        SimpleMailboxMessage message1InOtherMailbox = SimpleMailboxMessage.copy(benwaWorkMailbox.getMailboxId(), message1);
+        message1InOtherMailbox.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1InOtherMailbox);
+
+        List<MailboxId> mailboxes = sut.findMailboxes(message1.getMessageId());
+        assertThat(mailboxes).containsOnly(benwaInboxMailbox.getMailboxId(), benwaWorkMailbox.getMailboxId());
+    }
+
+    @ContractTest
+    public void saveShouldSaveAMessage() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1);
```

Comment:
Setting attributes on MailboxMessage should be done before, as MailboxMessage has to be seen as immutable.